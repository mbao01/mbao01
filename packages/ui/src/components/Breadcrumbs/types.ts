export type BreadcrumbProps = React.HTMLAttributes<HTMLLIElement>;

export type BreadcrumbsProps = React.HTMLAttributes<HTMLUListElement>;
