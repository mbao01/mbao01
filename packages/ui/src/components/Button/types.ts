import type { ButtonHTMLAttributes, ReactNode } from 'react';

export type ButtonSize = 'xs' | 'sm' | 'md' | 'lg';

export type ButtonVariant =
  | 'default'
  | 'neutral'
  | 'primary'
  | 'secondary'
  | 'accent'
  | 'ghost'
  | 'link'
  | 'info'
  | 'success'
  | 'warning'
  | 'error';

export type ButtonProps = {
  size?: ButtonSize;
  wide?: boolean;
  label: ReactNode;
  loading?: boolean;
  outline?: boolean;
  variant?: ButtonVariant;
  disabled?: boolean;
} & ButtonHTMLAttributes<HTMLButtonElement>;
