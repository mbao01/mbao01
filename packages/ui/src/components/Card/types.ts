import React from "react";

export type CardProps = {
  /**
   * applies smaller padding to the card.
   */
  compact?: boolean;
  /**
   * applies border to the card.
   */
  bordered?: boolean;
  /**
   * shows the card horizontally with image to the right or left of it.
   */
  horizontal?: boolean;
  /**
   * overlay is used to make the image full width with the content on top.
   */
  overlay?: boolean;
} & React.HTMLAttributes<HTMLDivElement>;

export type CardContentProps = React.HTMLAttributes<HTMLDivElement>;
export type CardImageProps = React.ImgHTMLAttributes<HTMLImageElement>;
export type CardHeaderProps = React.HTMLAttributes<HTMLDivElement>;
export type CardFooterProps = React.HTMLAttributes<HTMLDivElement>;
export type CardDescriptionProps = React.HTMLAttributes<HTMLParagraphElement>;
