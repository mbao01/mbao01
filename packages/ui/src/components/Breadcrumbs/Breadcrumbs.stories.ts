import type { Meta, StoryObj } from "@storybook/react";
import {
  reactRouterParameters,
  withRouter,
} from "storybook-addon-react-router-v6";

import { Breadcrumbs } from "./Breadcrumbs";

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: Breadcrumbs,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
  decorators: [withRouter],
} satisfies Meta<typeof Breadcrumbs>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {
    root: "Home",
  },
};

export const EditProfilePage: Story = {
  parameters: {
    reactRouter: reactRouterParameters({
      routing: { path: "/profile/edit" },
    }),
  },
  args: {
    root: "Dashboard",
  },
};
