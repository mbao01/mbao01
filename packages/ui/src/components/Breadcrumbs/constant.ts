export const getBreadcrumbs = (
  pathname: string,
  labels?: Record<string, string>
) => {
  const segments = pathname.split("/").filter(Boolean);

  const breadcrumbs = segments.map((segment, index) => {
    const path = `/${segments.slice(0, index + 1).join("/")}`;
    return {
      href: { pathname: path },
      segment: labels?.[segment] || segment,
    };
  });

  return breadcrumbs;
};
