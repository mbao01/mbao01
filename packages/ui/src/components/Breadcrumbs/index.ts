export { Breadcrumbs } from './Breadcrumbs';
export { getSubpaths } from "./getSubpaths";
