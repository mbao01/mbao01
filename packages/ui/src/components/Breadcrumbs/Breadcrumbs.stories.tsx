import type { Meta, StoryContext, StoryObj } from "@storybook/react";
import { Breadcrumb, Breadcrumbs } from "./Breadcrumbs";
import { getSubpaths } from ".";

type ArgsType = {
  pathname: string;
  subpathLabels?: Record<string, string>;
  includeRoot?: boolean;
};

const withBreadcrumbs = (_: any, context: StoryContext<ArgsType>) => {
  const { pathname, subpathLabels, includeRoot } = context.args;
  const subpaths = getSubpaths(pathname, subpathLabels, includeRoot);

  return (
    <Breadcrumbs>
      {subpaths.map((subpath) => (
        <Breadcrumb>
          <a href={subpath.href.pathname}>{subpath.label}</a>
        </Breadcrumb>
      ))}
    </Breadcrumbs>
  );
};

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  title: "Components/Breadcrumbs",
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  args: {
    pathname: "/profile/edit",
    subpathLabels: {
      "/": "Dashboard",
    },
    includeRoot: true,
  },
  argTypes: {
    pathname: {
      control: "text",
      description: "URL pathname",
    },
    subpathLabels: {
      control: "object",
      description: "Object mapping a subpath to a label",
    },
    includeRoot: {
      control: "boolean",
      description: "Include the root path or not",
    },
  },
  decorators: [withBreadcrumbs],
} satisfies Meta<ArgsType>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: { pathname: "/" },
};

export const WithSubpathLabels: Story = {
  args: {
    subpathLabels: {
      "/": "Dashboard",
      "/profile": "Profile",
      "/profile/edit": "Edit Profile",
    },
  },
};

export const WithoutSubpathLabels: Story = {
  args: {
    pathname: "/profile/edit",
    includeRoot: false,
  },
};
