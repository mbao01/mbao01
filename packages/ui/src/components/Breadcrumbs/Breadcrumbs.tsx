import { useLocation } from "react-router-dom";
import { getBreadcrumbs } from "./constant";
import type { BreadcrumbsProps } from "./types";
import { Link } from "../Link";

export const Breadcrumbs = ({ root, labels }: BreadcrumbsProps) => {
  const location = useLocation();
  const breadcrumbs = getBreadcrumbs(location.pathname, labels);

  return (
    <div className="breadcrumbs -ml-1 w-fit px-1 text-sm">
      <ul>
        {root && (
          <li>
            <Link hover href="/">
              {root}
            </Link>
          </li>
        )}
        {breadcrumbs.map((crumb) => (
          <li key={crumb.href.pathname}>
            <Link hover href={crumb.href} className="capitalize">
              {crumb.segment}
            </Link>
          </li>
        ))}
      </ul>
    </div>
  );
};
