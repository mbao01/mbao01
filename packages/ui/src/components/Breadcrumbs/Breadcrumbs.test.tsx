import { render, screen, within } from "@testing-library/react";
import { MemoryRouter, type MemoryRouterProps } from "react-router-dom";
import { Breadcrumbs } from "./Breadcrumbs";
import { type BreadcrumbsProps } from "./types";

describe("Breadcrumbs", () => {
  const renderBreadcrumbs = (
    props?: BreadcrumbsProps,
    memoryRouterProps?: MemoryRouterProps
  ) => {
    return render(
      <MemoryRouter {...memoryRouterProps}>
        <Breadcrumbs {...props} />
      </MemoryRouter>
    );
  };

  it("renders empty list", () => {
    const { asFragment } = renderBreadcrumbs();

    expect(screen.getByRole("list")).toBeEmptyDOMElement();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders root path", () => {
    const { asFragment } = renderBreadcrumbs({ root: "Home" });

    const listItemEl = within(screen.getByRole("listitem"));
    const anchorEl = listItemEl.getByRole("link", { name: "Home" });

    expect(anchorEl).toBeInTheDocument();
    expect(anchorEl).toHaveAttribute("href", "/");
    expect(asFragment()).toMatchSnapshot();
  });

  it("shows unlabeled paths", () => {
    const { asFragment } = renderBreadcrumbs(
      { root: "dashboard" },
      { initialEntries: ["/profile/edit"] }
    );

    [
      { name: "dashboard", href: "/" },
      { name: "profile", href: "/profile" },
      { name: "edit", href: "/profile/edit" },
    ].forEach(({ name, href }) => {
      const anchorEl = screen.getByRole("link", { name });

      expect(anchorEl).toBeInTheDocument();
      expect(anchorEl).toHaveAttribute("href", href);
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("shows labeled paths", () => {
    const { asFragment } = renderBreadcrumbs(
      { labels: { profile: "My Profile", edit: "Edit Profile" } },
      { initialEntries: ["/profile/edit"] }
    );

    [
      { name: "My Profile", href: "/profile" },
      { name: "Edit Profile", href: "/profile/edit" },
    ].forEach(({ name, href }) => {
      const anchorEl = screen.getByRole("link", { name });

      expect(anchorEl).toBeInTheDocument();
      expect(anchorEl).toHaveAttribute("href", href);
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
