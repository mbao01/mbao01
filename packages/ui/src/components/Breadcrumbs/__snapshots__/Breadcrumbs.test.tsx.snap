// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Breadcrumbs > renders empty list 1`] = `
<DocumentFragment>
  <div
    class="breadcrumbs -ml-1 w-fit px-1 text-sm"
  >
    <ul />
  </div>
</DocumentFragment>
`;

exports[`Breadcrumbs > renders root path 1`] = `
<DocumentFragment>
  <div
    class="breadcrumbs -ml-1 w-fit px-1 text-sm"
  >
    <ul>
      <li>
        <a
          href="/"
        >
          Home
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Breadcrumbs > shows labeled paths 1`] = `
<DocumentFragment>
  <div
    class="breadcrumbs -ml-1 w-fit px-1 text-sm"
  >
    <ul>
      <li>
        <a
          href="/profile"
        >
          My Profile
        </a>
      </li>
      <li>
        <a
          href="/profile/edit"
        >
          Edit Profile
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Breadcrumbs > shows unlabeled paths 1`] = `
<DocumentFragment>
  <div
    class="breadcrumbs -ml-1 w-fit px-1 text-sm"
  >
    <ul>
      <li>
        <a
          href="/"
        >
          dashboard
        </a>
      </li>
      <li>
        <a
          href="/profile"
        >
          profile
        </a>
      </li>
      <li>
        <a
          href="/profile/edit"
        >
          edit
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
